File: entertainmentbox/entertainmentbox.py
#!/usr/bin/env python

import sys
from collections import namedtuple

Show = namedtuple('Show', ['start', 'stop'])
Recorder = namedtuple('Recorder', ['id', 'end'])

def entertain_new(k, shows):
    count = 0
    shows = sorted(shows, key=lambda x: x.stop)
    recorders = [ Recorder(i, 0) for i in range(k) ]

    for show in shows:
        recorder = reduce(lambda best, rec: rec if best.end > show.start and rec.end <= show.start or rec.end > best.end and rec.end <= show.start else best, recorders, recorders[0])
        if recorder.end <= show.start:
            recorders[recorder.id] = Recorder(recorder.id, show.stop)
            count += 1

    return count

def entertain(k, shows):
    count = 0
    shows = sorted(shows, key=lambda x: x.stop)
    recorders = [ 0 ] * k
    ind = 0

    t = shows[0].start

    while ind < len(shows):
        for i in range(len(recorders)):
            rec = recorders[i]
            if rec <= t:
                recorders[i] = 0

        show = shows[ind]
        if show.stop <= t:
            # show has ended already
            ind += 1
        elif show.start == t:
            ind += 1
            try:
                i = recorders.index(0)
                recorders[i] = show.stop
                count += 1
                #print 'Choosing job {j} on recorder {i}'.format(j=show, i=i)
            except:
                #print 'Discarding job {j}'.format(j=show)
                pass

        if ind < len(shows):
            t = shows[ind].start
        #print 'New time: {t}'.format(t=t)
    return count


def main():
    _, k = [int(x) for x in raw_input().split()]
    shows = []
    for line in sys.stdin:
        start, stop = [int(x) for x in line.split()]
        shows.append(Show(start,stop))

    print entertain_new(k, shows)

if __name__ == '__main__':
	sys.exit(main())

File: entertainmentbox/entertainmentbox.cc
#include <iostream>
#include <set>
#include <numeric>

using namespace std;

class Show {
	public:
		int start;
		int stop;
		int id;

		Show(int start, int stop, int id) {
			this->start = start;
			this->stop = stop;
			this->id = id;
		}

		bool operator<(const Show& rhs) const {
			return this->stop == rhs.stop ? this->id < rhs.id : this->stop < rhs.stop;
		}

		bool operator>(const Show& rhs) const {
			return rhs < *this;
		}
};

ostream& operator<<(ostream& os, const Show& g) {
	return os << "[" << g.start << ", " << g.stop << "]: " << g.id;
}

class Recorder {
	public:
		int end;

		Recorder() : end(0) { }
};

int main(int argc, const char *argv[])
{
	int n, k;
	cin >> n >> k;
	set<Show> shows;
	Recorder *recorders = new Recorder[k];

	while (n) {
		int start, stop;
		cin >> start >> stop;
		Show show(start, stop, n);
		shows.insert(show);
		n--;
	}

	int count = 0;

	for (set<Show>::iterator iter = shows.begin(); iter != shows.end(); ++iter) {
		const Show &show = *iter;
		auto tmp = [show](Recorder *best, Recorder &rec) -> Recorder* {
			if ((best->end > show.start && rec.end <= show.start) || (rec.end > best->end && rec.end <= show.start)) {
				return &rec;
			} else {
				return best;
			}
		};

		Recorder *best = accumulate(recorders, recorders + k, recorders, tmp);
		if (best->end <= show.start) {
			best->end = show.stop;
			++count;
		}
	}

	cout << count << endl;

	return 0;
}
